### PR TITLE
[C-IRIS] move cspace_free_polytope.h out of dev folder.

### PR DESCRIPTION
**Pull request by hongkai-dai**
I will move the cc file out of dev folder in my next PR.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18860)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@EricCousineau-TRI for platform review please, thanks!

Could you platform review the file cspace_free_polytope.h? It lived in the dev folder before and has been feature reviewed already.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-:-NOziQAs4TFjHc5tl6kj:b-tci7qo)* status: LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @EricCousineau-TRI)

**Review by EricCousineau-TRI — COMMENTED** [review]
Looks good thus far, though I'm a bit sketched out by the boundary cross of "non-dev header and dev source file"

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-:-NP-8vyu2ESqAPmzWzk6:b-qeermw)* status: 6 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NOzqRvg5i36eg0suEEW):*
Why not include the cc file in this PR? I'd be happy to review it that way.

I'm not sure if I understand the benefit from separating 'em out, esp. given the Bazel oddities below (referring to direct source file in separate dir).
___
*[`geometry/optimization/cspace_free_polytope.h` line 44 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NOzqcxN70Jh1AKTyfNK:-NOzqcxN70Jh1AKTyfNL:bom3423) ([raw file](https://github.com/RobotLocomotion/drake/blob/0d405237a5f6deabfd73b1cae5035b6dd033ff53/geometry/optimization/cspace_free_polytope.h#L44)):*
> ```objc
> 
> /**
>  This class tries to find large convex polytope in the tangential-configuration
> ```

nit grammar "find a"
___
*[`geometry/optimization/cspace_free_polytope.h` line 156 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NP-76ND0SI3q0IFLlQt:-NP-76ND0SI3q0IFLlQu:b-jcioq6) ([raw file](https://github.com/RobotLocomotion/drake/blob/0d405237a5f6deabfd73b1cae5035b6dd033ff53/geometry/optimization/cspace_free_polytope.h#L156)):*
> ```objc
>    When searching for the separating plane, we want to certify that the
>    numerator of a rational is non-negative in the C-space region C*s<=d,
>    s_lower<= s <=s_upper. Hence for each of the rational we will introduce
> ```

nit Missing space between `s_lower<=`
___
*[`geometry/optimization/cspace_free_polytope.h` line 176 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NP-7OWD9lSjIfEu2ui8:-NP-7OWD9lSjIfEu2ui9:b-up1a8f) ([raw file](https://github.com/RobotLocomotion/drake/blob/0d405237a5f6deabfd73b1cae5035b6dd033ff53/geometry/optimization/cspace_free_polytope.h#L176)):*
> ```objc
>   /**
>    We certify that a pair of geometries is collision free in the C-space region
>    {s | Cs<=d, s_lower<=s<=s_upper}, by finding the separating plane and the
> ```

nit Extra commas `,`?
___
*[`geometry/optimization/cspace_free_polytope.h` line 434 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NP-88SN7nra9cfNA-eC:-NP-88SN7nra9cfNA-eD:bf7kerj) ([raw file](https://github.com/RobotLocomotion/drake/blob/0d405237a5f6deabfd73b1cae5035b6dd033ff53/geometry/optimization/cspace_free_polytope.h#L434)):*
> ```objc
>   /**
>    Constructs the MathematicalProgram which searches for a separation
>    certificate for a pair of geometries for a C-space polytope.Search for the
> ```

nit missing Space?
___
*[`geometry/optimization/cspace_free_polytope.h` line 577 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NP-8dCr5SxKH78WWPZk:-NP-8dCr5SxKH78WWPZl:b-ct5b54) ([raw file](https://github.com/RobotLocomotion/drake/blob/0d405237a5f6deabfd73b1cae5035b6dd033ff53/geometry/optimization/cspace_free_polytope.h#L577)):*
> ```objc
>   /** Adds the constraint that the ellipsoid {Q*u+s₀ | uᵀu≤1} is inside the
>      polytope {s | C*s <= d} with margin δ. Namely for the i'th face cᵢᵀs≤dᵢ, we
>      have |cᵢᵀQ|₂ ≤ dᵢ − cᵢᵀs₀ − δᵢ and |cᵢ|₂≤1
> ```

Unclear what `c1, d1, and del1` are w.r.t. arguments.

**Review by hongkai-dai — COMMENTED** [review]
Thanks Eric. The source file is 1800 lines by itself. I don't want to put too much burden on the reviewer. But if you are OK with it, I can include that in this PR also. I will also look for a second platform reviewer, according to the rule https://drake.mit.edu/developers.html#review-process:~:text=This%20rule%20may%20be%20overridden%20by%20agreement%20of%20at%20least%20two%20platform%20reviewers%20(listed%20below).

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-:-NP-EW7y5L-JLI4OyTfr:bsas6v1)* status: 6 unresolved discussions, LGTM missing from assignee EricCousineau-TRI(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

**Review by EricCousineau-TRI — COMMENTED** [review]
Ah, thank you for pointing that out! Didn't realize full extent of lines.
Yes, let's land this first just to keep things separated out.

:lgtm: x2 platform +(status: single reviewer ok) +(release notes: feature) 

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-:-NP-dwkV5xt9VBDpLQR3:bedra6q)* status: :shipit: complete! all discussions resolved, LGTM from assignee EricCousineau-TRI(platform) (waiting on @hongkai-dai)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/18860#-NOzqRvg5i36eg0suEEW):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Why not include the cc file in this PR? I'd be happy to review it that way.

I'm not sure if I understand the benefit from separating 'em out, esp. given the Bazel oddities below (referring to direct source file in separate dir).
</blockquote></details>

OK Per above, it seems to be simplest to merge forward.

**Comment by jwnimmer-tri** [comment]
> > > Looks good thus far, though I'm a bit sketched out by the boundary cross of "non-dev header and dev source file"
> >
> > The source file is 1800 lines by itself. I don't want to put too much burden on the reviewer.
>
> Yes, let's land this first just to keep things separated out.

This was a procedural mistake.

For the past several months since this landed, we have been installing the header file and its machine code as part of stable releases, for users to call at will, even thought the cc file that compiles down into the machine code was never reviewed.

The better procedure would have been to platform review the header file, but not move it out of `dev` yet.  Either in an issue checklist or comments in the code, we could keep track that its been reviewed.  Likewise we can do the cc file review without moving it.  Then in one swoop we move the already-reviewed files up a level, without re-reviewing them in total.

Alternatively, we could have moved the header out of dev like we did here, but mark it private (`internal = True` in BUILD) and not install it, until the cc file was reviewed as well.

In short -- the cardinal rule remains intact: never install `dev` code.

File: geometry/optimization/dev/BUILD.bazel [hunk 1]
```diff
@@ -12,20 +12,7 @@ load(
...
-        "//common/symbolic:monomial_util",
-        "//geometry/optimization:c_iris_collision_geometry",
-        "//geometry/optimization:c_iris_separating_plane",
-        "//solvers:choose_best_solver",
-        "//solvers:mathematical_program",
-        "//solvers:mosek_solver",
-        "//solvers:solve",
-    ],
-)
+exports_files(["cspace_free_polytope.cc"])
```
Comment (jwnimmer-tri) [inline geometry/optimization/dev/BUILD.bazel:15]: FYI This is where things went wrong; a `dev/BUILD.bazel` file must always have narrow visibility.  Exporting to `visibility:public` broke through the firewall.
